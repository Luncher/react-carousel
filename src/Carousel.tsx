import React, { Children, useState, useCallback, useEffect, cloneElement, isValidElement, ReactNode } from "react"
import cns from 'classnames'

import './carousel.css'

export type DotPosition = 'left' | 'right' | 'top' | 'bottom'

export type EasingType = 'linear'

export type EffectType = 'scroll' | 'fade'

export interface Props {
  autoplay?: boolean
  dotPosition?: DotPosition
  easing?: EasingType
  effect?: EffectType
  interval?: number
  afterChange?: (current: number) => void
  beforeChange?: (from: number, to: number) => void
  children: ReactNode
}

export default function Carousel({
  autoplay,
  interval = 1000,
  children,
  dotPosition,
  beforeChange,
  afterChange
}: Props) {
  const [position, setPosition] = useState({
    prev: 0,
    cur: 0
  })

  const cnt = Children.map(children, (it: ReactNode, index) => {
    if (!isValidElement(it)) {
      console.warn('Carousel children must be an valid element')
      return null
    }
    const classNames = cns(it.props.className, 'carousel-item', {
      'carousel-item-current': index === position.cur,
      'carousel-item-prev': index !== position.cur && index === position.prev
    })
    return cloneElement(it, { className: classNames })
  })

  const handleChange = useCallback(
    (value: number) => {
      beforeChange?.(position.cur, value)
      setPosition({ prev: position.cur, cur: value })
    },
    [position, beforeChange]
  )

  useEffect(() => {
    afterChange?.(position.cur)
  }, [position.cur, afterChange])

  if (!cnt || !cnt.length) {
    return null
  }

  return (
    <div className="carousel">
      <div className="carousel-inner">
        {cnt}
      </div>
      <CarouselDot
        value={position.cur}
        onChange={handleChange}
        position={dotPosition}
        dotCount={cnt.length}
      />
    </div>
  )
}


interface CarouselDotProps {
  position?: DotPosition
  dotCount: number
  value: number
  onChange: (value: number) => void
}

function CarouselDot({
  value,
  onChange,
  dotCount,
  position = 'bottom'
}: CarouselDotProps) {

  const classNames = cns(
    'carousel-dot',
    `carousel-dot-${position}`
  )

  const cnt = []
  for (let i = 0; i < dotCount; i++) {
    const itemClassNames = cns({
      'carousel-dot-item': true,
      'carousel-dot-item-active': i === value
    })
    cnt.push(
      <p className={itemClassNames} onClick={() => onChange(i)}></p>
    )
  }

  return (
    <div className={classNames}>
      {cnt}
    </div>
  )
}
