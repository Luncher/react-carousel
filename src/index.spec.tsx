import React, { CSSProperties } from 'react'
import renderer from 'react-test-renderer'

import Carousel from './index'

const style: CSSProperties = {
  lineHeight: '150px',
  textAlign: 'center',
  background: '#364d79',
  color: '#fff',
  margin: 0,
}

it('renders correctly', () => {
  const tree = renderer
    .create(
      <Carousel>
        <div>
          <p style={style}>1</p>
        </div>
        <div>
          <p style={style}>2</p>
        </div>
        <div>
          <p style={style}>3</p>
        </div>
      </Carousel>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('renders autoplay, effect, dotPosition correctly', () => {
  const tree = renderer
    .create(
      <Carousel autoplay effect="fade" dotPosition="left">
        <div>
          <p style={style}>1</p>
        </div>
        <div>
          <p style={style}>2</p>
        </div>
        <div>
          <p style={style}>3</p>
        </div>
      </Carousel>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})
