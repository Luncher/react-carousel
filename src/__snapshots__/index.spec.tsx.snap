// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders autoplay correctly 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-inner"
  >
    <div
      className="carousel-item carousel-item-initial"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        1
      </p>
    </div>
    <div
      className="carousel-item"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        2
      </p>
    </div>
    <div
      className="carousel-item"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        3
      </p>
    </div>
  </div>
  <div
    className="carousel-dot carousel-dot-bottom"
  >
    <p
      className="carousel-dot-item carousel-dot-item-active"
      onClick={[Function]}
    />
    <p
      className="carousel-dot-item"
      onClick={[Function]}
    />
    <p
      className="carousel-dot-item"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`renders autoplay, effect, dotPosition correctly 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-inner"
  >
    <div
      className="carousel-item carousel-item-initial"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        1
      </p>
    </div>
    <div
      className="carousel-item"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        2
      </p>
    </div>
    <div
      className="carousel-item"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        3
      </p>
    </div>
  </div>
  <div
    className="carousel-dot carousel-dot-left"
  >
    <p
      className="carousel-dot-item carousel-dot-item-active"
      onClick={[Function]}
    />
    <p
      className="carousel-dot-item"
      onClick={[Function]}
    />
    <p
      className="carousel-dot-item"
      onClick={[Function]}
    />
  </div>
</div>
`;

exports[`renders correctly 1`] = `
<div
  className="carousel"
>
  <div
    className="carousel-inner"
  >
    <div
      className="carousel-item carousel-item-initial"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        1
      </p>
    </div>
    <div
      className="carousel-item"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        2
      </p>
    </div>
    <div
      className="carousel-item"
      style={
        Object {
          "animationTimingFunction": "linear",
        }
      }
    >
      <p
        style={
          Object {
            "background": "#364d79",
            "color": "#fff",
            "lineHeight": "150px",
            "margin": 0,
            "textAlign": "center",
          }
        }
      >
        3
      </p>
    </div>
  </div>
  <div
    className="carousel-dot carousel-dot-bottom"
  >
    <p
      className="carousel-dot-item carousel-dot-item-active"
      onClick={[Function]}
    />
    <p
      className="carousel-dot-item"
      onClick={[Function]}
    />
    <p
      className="carousel-dot-item"
      onClick={[Function]}
    />
  </div>
</div>
`;
